### ISSUE TITLE
more specifically API

### ISSUE THREAD

**Issue by Kozlov-V** [titan098/rackspace_cloud_files#1]
Hi
I would like to ask you help and tell more about more specifically the less i m need the following API:
- download the file in the CDN;
- install the storage time of a file (for example 1 month)
- to give the file the user through the stream of link TempURL.

I'm very grateful for your help.

Thank you!

**Comment by titan098** [comment]
Hello,

I have added some functions that will hopefully address your comments. I used the functions provided in the `rackspace_cloud_files` module, as I imagine that this is how most people will interact with the API. The functions can be used directly from the `rackspace_cloud_files_api` module, however the state for the connection would have to be maintained manually.

I hope that this is useful to you.

**To download a file in the CDN you should first enable the CDN for a container**
1. Enable the CDN for a container in Cloud Files
2. `rackspace_cloud_files:cdn_get_object_url/3` can be used to return the URL for a object in the CDN-enabled container.
3. The URL type that can be passed is either `url`, `ssl_url`, or `streaming_url`, which correspond to the URL types returned by Cloud Files.

``` erlang
1> application:start(rackspace_cloud_files).
ok
2> rackspace_cloud_files:get_auth_token(uk, Username, APIKey).
ok
3> rackspace_cloud_files:list_containers().
["bob"]
4> rackspace_cloud_files:cdn_list_container("true").
[]
5> rackspace_cloud_files:create_object("bob", "hello_world.txt", <<"Hello World">>).
ok
6> rackspace_cloud_files:cdn_enable("bob", 900, false).
[{ssl_url,"https://5211286d8f307bf674cc-47afd4e8fb7ffb0d393f21f19dc43cfd.ssl.cf3.rackcdn.com"},
 {url,"http://0e9f7bb7c9681343e4af-47afd4e8fb7ffb0d393f21f19dc43cfd.r58.cf3.rackcdn.com"},
 {streaming_url,"http://6726f1a8e2dfcba31e6e-47afd4e8fb7ffb0d393f21f19dc43cfd.r58.stream.cf3.rackcdn.com"}]
7> rackspace_cloud_files:list_objects("bob").
["hello_world.txt"]
8> rackspace_cloud_files:cdn_get_object_url(ssl_url, "bob", "hello_world.txt").
"https://5211286d8f307bf674cc-47afd4e8fb7ffb0d393f21f19dc43cfd.ssl.cf3.rackcdn.com/hello_world.txt"
9> ibrowse:send_req("https://5211286d8f307bf674cc-47afd4e8fb7ffb0d393f21f19dc43cfd.ssl.cf3.rackcdn.com/hello_world.txt", [], get).
{ok,"200",
    [{"Content-Length","11"},
     {"Accept-Ranges","bytes"},
     {"Last-Modified","Sun, 10 Mar 2013 15:27:15 GMT"},
     {"ETag","b10a8db164e0754105b7a99be72e3fe5"},
     {"X-Timestamp","1362929235.02291"},
     {"Content-Type","text/plain"},
     {"X-Trans-Id","txfc5c8996c3024594bb0dc9b20f872376"},
     {"Cache-Control","public, max-age=895"},
     {"Expires","Sun, 10 Mar 2013 15:42:20 GMT"},
     {"Date","Sun, 10 Mar 2013 15:27:25 GMT"},
     {"Connection","keep-alive"}],
    "Hello World"}
```

**To specify the delete time for an object**

You can either choose to specify the `X-Delete-At` or `X-Delete-After` headers when the object is created or for an existing object. The functions `rackspace_cloud_files:delete_object_at/3` and `rackspace_cloud_files:delete_object_after/3` have been added to modify an existing object's headers. Alternatively, `rackspace_cloud_files:create_object/5` can be used to specify the `X-Delete-` headers during object creation (also CORS options and the content type of the object).

_To specify the `X-Delete-` options at creation:_

``` erlang
2> rackspace_cloud_files:list_objects("bob").                                                  
[]
3> rackspace_cloud_files:create_object("bob", "hello_world.txt", <<"Hello World">>, [], [{delete_after, "900"}]).
ok
4> rackspace_cloud_files:retrieve_object_headers("bob", "hello_world.txt").                                      
[{"Content-Length","11"},
 {"X-Delete-At","1362929034"},
 {"Accept-Ranges","bytes"},
 {"Last-Modified","Sun, 10 Mar 2013 15:08:54 GMT"},
 {"Etag","b10a8db164e0754105b7a99be72e3fe5"},
 {"X-Timestamp","1362928134.12712"},
 {"Content-Type","text/plain"},
 {"X-Trans-Id","tx73cce58eb2ef48e9b2a8cc5035057bea"},
 {"Date","Sun, 10 Mar 2013 15:09:18 GMT"}]
```

_To modify the `X-Delete-` options for an existing object:_

``` erlang
6> rackspace_cloud_files:list_objects("test").
["test.txt"]
7> rackspace_cloud_files:delete_object_after("test", "test.txt", 900).
ok
8> rackspace_cloud_files:retrieve_object_headers("test", "test.txt").
[{"Content-Length","28"},
 {"X-Delete-At","1362926104"},
 {"Accept-Ranges","bytes"},
 {"Last-Modified","Sun, 10 Mar 2013 14:20:04 GMT"},
 {"Etag","83d5db05f3f60e39b095847f151fc005"},
 {"X-Timestamp","1362925204.28738"},
 {"Content-Type","text/plain"},
 {"X-Trans-Id","tx6a7da3e1f10e48b2ab2946a289540bcf"},
 {"Date","Sun, 10 Mar 2013 14:25:11 GMT"}]
```

**To create and return a TempURL for an object**

The functions `rackspace_cloud_files:tempurl_set_key/1` and `rackspace_cloud_files:tempurl_create_url/6` where added to manage the creating of TempURLs. The functions can be used to pass a key to Cloud Files and then generate the correct HMAC-SHA1 signature for the requests. The functions allow allow `GET` and `PUT` URLs to be created. As per the Cloud Files API, if the key changes the TempURLs created with the old key will expire after 60 seconds, otherwise they will expire after the specified time.

``` erlang
2> rackspace_cloud_files:list_containers().
["test"]
3> rackspace_cloud_files:create_container("test2").
ok
4> rackspace_cloud_files:create_object("test2", "test2.txt", <<"Hello World">>).
ok
5> rackspace_cloud_files:tempurl_set_key("1234567890").
ok
6> rackspace_cloud_files:tempurl_create_url(get, "test2", "test2.txt", 120, "1234567890").
"https://storage101.lon3.clouddrive.com/v1/MossoCloudFS_d1bcd532-3cfd-4411-93aa-4e836c06d6f9/test2/test2.txt?temp_url_sig=1247d3528013404e0a87ba5690cf18882ba3a934&temp_url_expires=1362920254"
7> ibrowse:send_req("https://storage101.lon3.clouddrive.com/v1/MossoCloudFS_d1bcd532-3cfd-4411-93aa-4e836c06d6f8/test2/test2.txt?temp_url_sig=1247d3528013404e0a87ba5690cf18882ba3a934&temp_url_expires=1362920254", [], get).
{ok,"200",
    [{"Content-Length","11"},
     {"Accept-Ranges","bytes"},
     {"Last-Modified","Sun, 10 Mar 2013 12:54:45 GMT"},
     {"Etag","b10a8db164e0754105b7a99be72e3fe5"},
     {"X-Timestamp","1362920085.05312"},
     {"Content-Type","text/plain"},
     {"Content-Disposition","attachment; filename=test2.txt"},
     {"X-Trans-Id","txc4c83ad1db284791b6c2c3d6641e8854"},
     {"Date","Sun, 10 Mar 2013 12:55:59 GMT"}],
    "Hello World"}
8> rackspace_cloud_files:list_objects("test2").
["test2.txt"]
9> rackspace_cloud_files:tempurl_create_url(put, "test2", "newfile.txt", 120, "1234567890").
"https://storage101.lon3.clouddrive.com/v1/MossoCloudFS_d1bcd532-3cfd-4411-93aa-4e836c06d6f4/test2/newfile.txt?temp_url_sig=770523720baa8999fa06a8d6f1f8a83f11253e19&temp_url_expires=1362920451"
10> ibrowse:send_req("https://storage101.lon3.clouddrive.com/v1/MossoCloudFS_d1bcd532-3cfd-4411-93aa-4e836c06d6f8/test2/newfile.txt?temp_url_sig=770523720baa8999fa06a8d6f1f8a83f11253e19&temp_url_expires=1362920451", [], put, <<"Hello World Again">>).
{ok,"201",
    [{"Last-Modified","Sun, 10 Mar 2013 12:59:51 GMT"},
     {"Content-Length","0"},
     {"Etag","7bac2e654b6275699acfad1ce68d9a5d"},
     {"Content-Type","text/html; charset=UTF-8"},
     {"X-Trans-Id","tx98af15b910424555b192cbc731f6a792"},
     {"Date","Sun, 10 Mar 2013 12:59:51 GMT"}],
    []}
11> rackspace_cloud_files:list_objects("test2").
["newfile.txt","test2.txt"]
12> rackspace_cloud_files:get_object("test2", "newfile.txt").
<<"Hello World Again">>
```

**Comment by Kozlov-V** [comment]
thank you very much for such quick work and all your help!

**Comment by Kozlov-V** [comment]
Hello!
I'm trying to learn your big decision, but not very good ...
here are my steps:

```
# root@s-nx rackspace_cloud_files/ebin# erl
Erlang R16B (erts-5.10.1) [source-05f1189] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V5.10.1  (abort with ^G)
1> rackspace_cloud_files:start().
{error,
    {bad_return,{{rackspace_cloud_files,start,[normal,[]]},ok}}}
=INFO REPORT==== 11-Mar-2013::13:09:01 ===
    application: rackspace_cloud_files
    exited: {bad_return,{{rackspace_cloud_files,start,[normal,[]]},ok}}
    type: temporary
2> application:start(rackspace_cloud_files).
=INFO REPORT==== 11-Mar-2013::13:10:02 ===
    application: rackspace_cloud_files
    exited: {bad_return,{{rackspace_cloud_files,start,[normal,[]]},ok}}
    type: temporary
{error,
    {bad_return,{{rackspace_cloud_files,start,[normal,[]]},ok}}}
```

i have error....

And so, when i try

```
1>rackspace_cloud_files:get_auth_token(uk, "myUserName", "99999999999999999999999")
```

i have:

```
** exception exit: {noproc,
                       {gen_server,call,
                           [rackspace_cloud_files_srv,
                            {get_auth_token,uk,"myUserName",
                                "99999999999999999999999"}]}}
     in function  gen_server:call/2 (gen_server.erl, line 180)
```

whether I use the "" (as string) in the Username and APIKey?
please help me, once again!

**Comment by titan098** [comment]
Hello,

The application has a dependency on [ibrowse](https://github.com/cmullaparthi/ibrowse) to perform the HTTP requests to the Cloud Files API (along with the Erlang `crypto`, `public_key`, and `ssl` modules).

`application:start(rackspace_cloud_files)` will try to start `ibrowse` and `rackspace_cloud_files_srv` in the supervisor `rackspace_cloud_files_sup`. If you are using [rebar](https://github.com/basho/rebar) to build the application in a larger package it should resolve the `ibrowse` dependency in the top-level deps directory.

The easiest way to start the application is from the top-level (not inside the ebin directory) with the command which will adjust the code search path to include the ebin directory for the application and the dependencies:

`erl -pa ebin -pa deps/*/ebin`

This will include the BEAM files for the `rackspace_cloud_files` application and any of the dependencies (only `ibrowse`).

Hope that this helps.

```
root# ./rebar get-deps && ./rebar compile
==> rackspace_cloud_files (get-deps)
Pulling ibrowse from {git,"https://github.com/cmullaparthi/ibrowse.git",
                          {branch,master}}
Cloning into 'ibrowse'...
==> ibrowse (get-deps)
==> ibrowse (compile)
Compiled src/ibrowse_app.erl
Compiled src/ibrowse.erl
Compiled src/ibrowse_http_client.erl
Compiled src/ibrowse_sup.erl
Compiled src/ibrowse_lb.erl
Compiled src/ibrowse_lib.erl
==> rackspace_cloud_files (compile)
Compiled src/rackspace_cloud_files_srv.erl
Compiled src/rackspace_cloud_files_sup.erl
Compiled src/rackspace_cloud_files.erl
Compiled src/rackspace_cloud_files_api.erl
root# erl -pa ebin -pa deps/*/ebin
Erlang R16B (erts-5.10.1) [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V5.10.1  (abort with ^G)
1> application:start(rackspace_cloud_files).
ok
2> rackspace_cloud_files:get_auth_token(uk, Username, APIKey).
ok
3> rackspace_cloud_files:list_containers().
[]
4> 
```

**Comment by Kozlov-V** [comment]
Voila! it's cool!)))
Earned!
I inserted  your project to my project based on a cowboy
And had the typical structure of the OTP applications.

```
myapp/
         /deps/
                /cowboy/
                /ranch/
                /ibrowse/
                /rackspace_cloud_files/
         /ebin/
         /src/
         rebar.config
         start.sh 
```

I did not understand what was my mistake. I removed all the DEPS.
Did get-deps and compile again and it worked!

AND also!!!!
For US-based accounts, need replaced in function:

```
get_auth_token(us, Username, APIKey) ->
    get_auth_token("https://identity.api.rackspacecloud.com/v1.0", Username, APIKey);
```

in modul rackspace_cloud_files_api.erl
About see at http://docs.rackspace.com/files/api/v1/cf-devguide/content/Authentication-d1e639.html#d6e626

Your app is outstanding!
Thanks for the help and support
